R/loo_moment_matching.R

```diff
@@ -0,0 +1,568 @@
+#' samples or not.
+#'   If \code{FALSE}, only the mean and marginal variances are matched.
+#' @template cores
+#' @param ... Further arguments passed to the custom functions documented above.
+#'
+#' @return The `mmloo()` methods return an updated \code{loo} object.
+#' The structure of the updated \code{loo} object is similar, but the
+#' method also stores the original Pareto k diagnostic values in
+#' the diagnostics field.
+#'
+#' @details The `mmloo()` function is an S3 generic and we provide a default
+#' method that takes as arguments user-specified functions \code{post_draws},
+#' \code{log_lik}, \code{unconstrain_pars}, \code{log_prob_upars}, and
+#' \code{log_lik_upars}.
+#'
+#' @section Defining `mmloo()` methods in a package: Package developers can
+#' define `mmloo()` methods for fitted models objects. The
+#' `mmloo.stanfit()` method in the **Examples** section provides an example.
+#'
+#' @seealso [loo()], [split_mmloo()]
+#' @template moment-matching-references
+#'
+#' @examples
+#' \dontrun{
+#' ### For package developers: defining mmloo methods
+#'
+#' # An example of a possible mmloo method for a 'stanfit' objects
+#' # (from package rstan). The example is just a wrapper of mmloo.stanfit
+#' # with user-defined functions post_draws_stanfit, log_lik_stanfit,
+#' # unconstrain_pars_stanfit, log_prob_upars_stanfit, and log_lik_upars_stanfit.
+#' #
+#' mmloo.stanfit <- function(x, loo, ...) {
+#' loo::mmloo.default(
+#'   x, loo = loo,
+#'   post_draws = post_draws_stanfit,
+#'   log_lik = log_lik_stanfit,
+#'   unconstrain_pars = unconstrain_pars_stanfit,
+#'   log_prob_upars = log_prob_upars_stanfit,
+#'   log_lik_upars = log_lik_upars_stanfit,
+#'   ...)
+#' }
+#' }
+#'
+#'
+
+
+
+
+#' @export
+mmloo <- function(x, ...) {
+  UseMethod("mmloo")
+}
+
+
+#' @describeIn mmloo A default method that takes as arguments
+#' a user-specified model object \code{x}, a \code{loo} object and
+#' user-specified functions \code{post_draws}, \code{log_lik},
+#' \code{unconstrain_pars}, \code{log_prob_upars}, and \code{log_lik_upars}.
+#' @export
+mmloo.default <- function(x, loo, post_draws, log_lik,
+                          unconstrain_pars, log_prob_upars,
+                          log_lik_upars, max_iters = 30L,
+                          k_thres = 0.5, split = TRUE,
+                          cov = TRUE, cores = getOption("mc.cores", 1),
+                          ...) {
+
+  # input checks
+  checkmate::assertClass(loo,classes = "loo")
+  checkmate::assertFunction(post_draws)
+  checkmate::assertFunction(log_lik)
+  checkmate::assertFunction(unconstrain_pars)
+  checkmate::assertFunction(log_prob_upars)
+  checkmate::assertFunction(log_lik_upars)
+  checkmate::assertNumber(max_iters)
+  checkmate::assertNumber(k_thres)
+  checkmate::assertLogical(split)
+  checkmate::assertLogical(cov)
+  checkmate::assertNumber(cores)
+
+
+  if ("psis_loo" %in% class(loo)) {
```

Comment:
Does it make sense to clean up the commented out code below?

Reply:
cleaned